core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2742,6 +2743,16 @@ private[spark] object Utils extends Logging {
     new File(path.getAbsolutePath + "." + UUID.randomUUID())
   }
 
+  /**
+   * Creates a file with group write permission.
+   */
+  def createFileAsGroupWritable(file: File): Unit = {
+    val perms = PosixFilePermissions.fromString("rw-rw----")
+    val path = file.toPath
+    Files.createFile(path)
+    Files.setPosixFilePermissions(path, perms)
```

Comment:
Yeah I saw that and wasn't sure what the background was that didn't work in the past.

Reply:
Based on the comment
```
* TODO: Find out why can't we create a dir using java api with permission 770
*  Files.createDirectories(mergeDir.toPath, PosixFilePermissions.asFileAttribute(
*  PosixFilePermissions.fromString("rwxrwx---")))
```

I'm thinking it might be a umask issue? The permissions you give it during dir creation will be affected by the umask, and the way around that is to create the dir and then change its permissions.

Reply:
IIRC this was not a umask issue (I had asked similar question to the team :-) ).
Will let @otterc comment more though - she has better context.

Reply:
Think I figured it out while testing this out in our env. Java has no way to set or keep the setgid bit when setting file permissions, so you lose the setgid bit with any of these methods of setting file permissions. I'm trying to figure out how yarn creates these folders since it manages to keep the bit, but it looks like it's doing it all through a local FileContext, which seems to be not keeping the setgid bit when making directories, so I'm not sure how that works.

Will keep investigating but will probably use the existing method to create the 770 folder

Reply:
Well sort of wrong again. At least in my environment, the setgid bit still gets removed when creating the dir with `mkdir -m`. I don't know if this is true on all linux systems, but it appears of the user trying to `chmod` a folder or file (even though the man page just says file), has to be in the group of the file in order to set or keep the setgid bit during a chmod call. Otherwise if you set the group write bit, it will remove the setgid bit. This caused all the shuffle blocks to not have the group of the yarn service (and therefore shuffle service), so the shuffle service couldn't read the shuffle files anymore.

After playing around, I was able to find a workaround by doing `umask 002 && mkdir <path>`, but that's getting more and more specific to linux systems. I can't think of any other way though, so definitely would love to hear some thoughts

Reply:
Thanks for the investigation @Kimahriman - @otterc can you comment on why we ended up needing this ? Thx

Reply:
We needed this for push-based shuffle to create merge directory for push-based shuffle. For push based shuffle, the shuffle service creates merged shuffle files under merge directory. These files can be read both by the shuffle server as well as the executor.
Yarn creates user dirs with the setgid bit [here](https://github.com/apache/hadoop/blob/7c97c0f969e6e04513e0b5920855e3b7095e8d2b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/container-executor/impl/container-executor.c#L1080).

Also, the container executor sets the umask to 0027 when launching the container [here](
https://github.com/apache/hadoop/blob/7c97c0f969e6e04513e0b5920855e3b7095e8d2b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/container-executor/impl/container-executor.c#L2435). So from the executor, just executing `mkdir <path>` should be sufficient.

Reply:
Thanks for pointing those links out. Is group write being `umask`'d out is what I'm trying to get around now. What I don't see is how the executor will be able to read the merged shuffle files in your case (which is sort of related to my problem but not fully).

The way I understand it based on what I've learned with Linux and the setgid stuff, let's say you're running your node manager and therefore shuffle service as `yarn:hadoop` like the docs suggest. And you're running your spark job as user `bob`.

- The DiskBlockManager creates the merge directory using `mkdir -p -m770 mergeDir/00` for each sub dir
- This would create `mergeDir` as `bob:hadoop` with mode `rwxr-s---` and `mergeDir/00` as `bob:hadoop` `rwxrwx---`. Notice that lost the setgid bit which is what I'm seeing in my environment (centos) and with a quick test in an ubuntu docker image. This is because `bob` is _not_ in the `hadoop` group so Linux removes the setgid bit when you change the group write bit
- The shuffle service will create files in `mergeDir/00` fine because it has group write permission, but the files would be created as `yarn:hadoop` with mode `rw-r-----`

So at that point it seems like the executor (`bob`) wouldn't be able to read the file (which would I assume only happen for locality purposes?)

Is there something I'm missing with that?

The issue I'm running into now is:
- The DiskBlockManager creates the blockmgr subdirs using `mkdir -p -m770 blockmgr/00`
- This creates the subdirs as `bob:hadoop` with mode `rwxrwx---` (no setgid bit anymore)
- Spark executor creates a shuffle file `blockmgr/00/shuffle_0_0_0.data` as `bob:bob` with mode `rw-r-----` because the parent dir didn't have the setgid bit anymore
- The shuffle service no longer has permission to read this file so it fails

The only ways around this I can think of are to try to `sh -c "umask 007 && mkdir -p <path"` when creating the subdirs, or to set a default group write facl on the blockmgr dir

Reply:
@Kimahriman I'll get back to you in a day or 2 regarding this.

Reply:
In our env the shuffle service is run as `yarn:yarn`.
Below is the explanation for push-based shuffle dirs:
> This would create mergeDir as bob:hadoop with mode rwxr-s--- and mergeDir/00 as bob:hadoop rwxrwx---. 

In our env, the merge directory and its subdirs are owned by `<user>:yarn` with permission `rwxrwx---`. Example:
`drwxrwx---  2 chsingh yarn 20480 Jan 20 22:38 merge_manager`

The merged files under the merge dirs are owned by `yarn:yarn` but have permissions `rw-rw-r--`. Example:
```
-rw-rw-r-- 1 yarn    yarn       40 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1252.index
-rw-rw-r-- 1 yarn    yarn       40 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1245.index
-rw-rw-r-- 1 yarn    yarn      632 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1208.meta
-rw-rw-r-- 1 yarn    yarn       40 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1208.index
-rw-rw-r-- 1 yarn    yarn       40 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1204.index
-rw-rw-r-- 1 yarn    yarn       40 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1200.index
-rw-rw-r-- 1 yarn    yarn      632 Jan 20 22:39 mergedShuffle_application_1632281309592_16957571_0_1262.meta
```
Since these files are created by shuffle service, they are created by the default permission. However these files are still not readable  by any user other than `chsingh` because the parent directory is not accessible by anyone else.

Reply:
Right right the shuffle service isn't affected by the 0027 umask so it's files world readable, that makes sense.

Doesn't help with this use case though 😭

Reply:
Also in that case was the reason for
```
*  TODO: Find out why can't we create a dir using java api with permission 770
*  Files.createDirectories(mergeDir.toPath, PosixFilePermissions.asFileAttribute(
*  PosixFilePermissions.fromString("rwxrwx---")))
```
actually just a umask issue that can be gotten around by create -> change permission in Java?

Reply:
No, I remember trying that but it didn't work with Java. It still created a directory with permission `750`

Reply:
Weird, it worked when I tried for the test case and when I tested it on my environment

Reply:
hmmm I think even @zhouyejoe tried that in our environment. @zhouyejoe do you remember whether that worked?

Reply:
I will give it a try again. Maybe made some mistake. Can't find any reason for it to not work. If it works, then we can clean it up.

Reply:
I was looking into this in our environment and was able to get the below piece of code working: 
`        if (dirToCreate.mkdirs() && dirToCreate.exists()) {
          Files.setPosixFilePermissions(dirToCreate.toPath,
            PosixFilePermissions.fromString("rwxrwx---"))
          created = dirToCreate
        }`
@Kimahriman do you mind testing this piece of code in your environment? It seems like the code referenced in the TODO does not pass the unit tests and also does not work in our environment.
If this looks good we can clean it up. 
cc: @zhouyejoe

Reply:
Yeah that's the umask issue. The todo does permission setting in the dir creation. If you `setPosixFilePermissions` after creating the directory you don't get limited by the umask (only weird behavior with the setgid bit mentioned in this thread). I've tried this approach in my env as well and it works in place of the createDirWith770 method